go/vt/vtgate/semantics/semantic_state.go

```diff
@@ -291,6 +291,14 @@ func (st *SemTable) AddExprs(tbl *sqlparser.AliasedTableExpr, cols sqlparser.Sel
 
 // TypeForExpr returns the type of expressions in the query
 func (st *SemTable) TypeForExpr(e sqlparser.Expr) (sqltypes.Type, collations.ID, bool) {
+	// We add a lot of WeightString() expressions to queries at late stages of the planning,
+	// which means that they don't have any type information. We can safely assume that they
+	// are VarBinary, since that's the only type that WeightString() can return.
+	_, isWS := e.(*sqlparser.WeightStringFuncExpr)
+	if isWS {
+		return sqltypes.VarBinary, collations.CollationBinaryID, true
+	}
+
 	if typ, found := st.ExprTypes[e]; found {
 		return typ.Type, typ.Collation, true
 	}
```

Comment:
should we rearrange this check and do this first 
```
if typ, found := st.ExprTypes[e]; found {
    return typ.Type, typ.Collation, true
}
```
if not found then check the expression for the weight_string function